File: benchmark/score.ts
/**
 * Score GAIA benchmark results.
 *
 * Reads results.jsonl, normalizes answers, computes accuracy.
 *
 * Usage: npx tsx benchmark/score.ts
 */
import fs from "node:fs";
import path from "node:path";

interface GaiaResult {
	task_id: string;
	expected_answer: string;
	model_answer: string | null;
	status: "success" | "error" | "no_answer";
	duration_ms: number;
	tokens?: { input: number; output: number; total: number };
	cost?: number;
	error?: string;
}

const BENCHMARK_DIR = path.resolve(import.meta.dirname ?? path.dirname(new URL(import.meta.url).pathname));
const RESULTS_FILE = path.join(BENCHMARK_DIR, "data", "results.jsonl");

function normalize(s: string): string {
	return s
		.toLowerCase()
		.trim()
		.replace(/\.$/, "") // trailing period
		.replace(/^["']|["']$/g, "") // surrounding quotes
		.replace(/^\$/, "") // leading $
		.replace(/,(\d{3})/g, "$1") // number commas: 1,000 -> 1000
		.trim();
}

function fuzzyMatch(expected: string, actual: string): boolean {
	const ne = normalize(expected);
	const na = normalize(actual);

	// Exact match
	if (ne === na) return true;

	// Numeric fuzzy match
	const numE = parseFloat(ne);
	const numA = parseFloat(na);
	if (!isNaN(numE) && !isNaN(numA)) {
		if (Math.abs(numE - numA) < 1e-6) return true;
		// Relative tolerance for larger numbers
		if (numE !== 0 && Math.abs((numE - numA) / numE) < 1e-6) return true;
	}

	return false;
}

function main() {
	if (!fs.existsSync(RESULTS_FILE)) {
		console.error("Results file not found at", RESULTS_FILE);
		console.error("Run 'npx tsx benchmark/run.ts' first.");
		process.exit(1);
	}

	const lines = fs.readFileSync(RESULTS_FILE, "utf-8").trim().split("\n");
	const results: GaiaResult[] = [];
	for (const line of lines) {
		if (!line) continue;
		try {
			results.push(JSON.parse(line));
		} catch {
			/* skip */
		}
	}

	let correct = 0;
	let answered = 0;
	let errors = 0;
	const wrong: Array<{ id: string; expected: string; got: string | null }> = [];
	let totalDuration = 0;
	let totalTokens = 0;
	let totalCost = 0;

	for (const r of results) {
		totalDuration += r.duration_ms;
		totalTokens += r.tokens?.total ?? 0;
		totalCost += r.cost ?? 0;

		if (r.status === "error") {
			errors++;
			wrong.push({ id: r.task_id.slice(0, 8), expected: r.expected_answer, got: r.error ?? "ERROR" });
			continue;
		}

		if (!r.model_answer) {
			wrong.push({ id: r.task_id.slice(0, 8), expected: r.expected_answer, got: null });
			continue;
		}

		answered++;
		if (fuzzyMatch(r.expected_answer, r.model_answer)) {
			correct++;
		} else {
			wrong.push({ id: r.task_id.slice(0, 8), expected: r.expected_answer, got: r.model_answer });
		}
	}

	const total = results.length;
	const accuracy = total > 0 ? ((correct / total) * 100).toFixed(1) : "0.0";
	const avgDuration = total > 0 ? Math.round(totalDuration / total / 1000) : 0;
	const avgTokens = total > 0 ? Math.round(totalTokens / total) : 0;

	console.log("=== GAIA Level 2 Benchmark Results ===\n");
	console.log(`Total tasks:    ${total}`);
	console.log(`Answered:       ${answered}`);
	console.log(`Errors:         ${errors}`);
	console.log(`No answer:      ${total - answered - errors}`);
	console.log(`Correct:        ${correct}`);
	console.log(`Accuracy:       ${accuracy}%`);
	console.log(`Avg duration:   ${avgDuration}s`);
	console.log(`Avg tokens:     ${avgTokens}`);
	console.log(`Total cost:     $${totalCost.toFixed(2)}`);

	if (wrong.length > 0) {
		console.log(`\n=== Wrong/Missing Answers (${wrong.length}) ===\n`);
		for (const w of wrong) {
			console.log(`  ${w.id}  expected: ${w.expected}  got: ${w.got ?? "NO ANSWER"}`);
		}
	}
}

main();

File: python/benchmark/score.py
"""GAIA benchmark scoring — compare model answers to expected answers."""

from __future__ import annotations

import json
import re
import sys
from pathlib import Path


def normalize_answer(answer: str) -> str:
    """Normalize an answer for comparison."""
    answer = answer.strip()
    # Remove trailing punctuation
    answer = answer.rstrip(".")
    # Lowercase
    answer = answer.lower()
    # Normalize whitespace
    answer = re.sub(r"\s+", " ", answer)
    return answer


def answers_match(expected: str, actual: str) -> bool:
    """Check if two answers match (case-insensitive, whitespace-normalized)."""
    norm_expected = normalize_answer(expected)
    norm_actual = normalize_answer(actual)

    # Exact match after normalization
    if norm_expected == norm_actual:
        return True

    # Try numeric comparison
    try:
        exp_num = float(norm_expected.replace(",", ""))
        act_num = float(norm_actual.replace(",", ""))
        if abs(exp_num - act_num) < 1e-6:
            return True
    except (ValueError, TypeError):
        pass

    # Check if expected is contained in actual (for longer-form answers)
    if norm_expected in norm_actual:
        return True

    return False


def score_results(results_file: str) -> dict:
    """Score results from a JSONL file."""
    results = []
    with open(results_file, "r") as f:
        for line in f:
            line = line.strip()
            if line:
                results.append(json.loads(line))

    total = len(results)
    correct = 0
    errors = 0
    timeouts = 0
    incorrect = []

    for r in results:
        status = r.get("status", "")
        expected = r.get("expected_answer", "")
        actual = r.get("model_answer", "")

        if status == "error":
            errors += 1
            continue
        if status == "timeout":
            timeouts += 1
            continue

        if answers_match(expected, actual):
            correct += 1
        else:
            incorrect.append({
                "task_id": r.get("task_id", ""),
                "expected": expected,
                "actual": actual,
            })

    accuracy = correct / total if total > 0 else 0

    return {
        "total": total,
        "correct": correct,
        "incorrect": len(incorrect),
        "errors": errors,
        "timeouts": timeouts,
        "accuracy": accuracy,
        "incorrect_details": incorrect,
    }


def main() -> None:
    """Score results and print summary."""
    from benchmark.config import RESULTS_FILE

    results_file = sys.argv[1] if len(sys.argv) > 1 else RESULTS_FILE

    if not Path(results_file).exists():
        print(f"Results file not found: {results_file}")
        sys.exit(1)

    scores = score_results(results_file)

    print(f"\n{'='*50}")
    print(f"GAIA Benchmark Results")
    print(f"{'='*50}")
    print(f"Total tasks:    {scores['total']}")
    print(f"Correct:        {scores['correct']}")
    print(f"Incorrect:      {scores['incorrect']}")
    print(f"Errors:         {scores['errors']}")
    print(f"Timeouts:       {scores['timeouts']}")
    print(f"Accuracy:       {scores['accuracy']:.1%}")
    print(f"{'='*50}")

    if scores["incorrect_details"]:
        print(f"\nIncorrect answers:")
        for item in scores["incorrect_details"]:
            print(f"  Task {item['task_id'][:8]}...")
            print(f"    Expected: {item['expected']}")
            print(f"    Got:      {item['actual']}")


if __name__ == "__main__":
    main()
